tc/core/tensor_api_rfc.h

```diff
@@ -0,0 +1,166 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include <dlpack/dlpack.h>
+
+#include "tc/proto/compcache.pb.h"
+
+/**
+ * Various tensor utilities used in the Tensor Comprehensions compiler.
+ * At a high-level we use the DLTensor type from dlpack and we add a
+ * DLConstTensor to propagate const-correctness.
+ * DLTensor is used as an ML-framework and backend-agnostic representation to
+ * interact with user-level tensors.
+ * DLTensorUPtr and DLConstTensorUPtr are used as metadata-owning pointers.
+ * The underlying concrete device pointer is borrowed when calling
+ * ExecutionEngine::compile and ExecutionEngine::run.
+ *
+ * Internally, TC uses TensorInfo for compiling and interacting with the
+ * compilation caches. TensorInfo is backed by protobuf and stored directly in
+ * the caches.
+ */
+#ifdef __cplusplus
+extern "C" {
+#endif
+/**
+ * DLPack is missing a ConstTensor. Add it here for now
+ */
+typedef struct {
+  const void* data;
+  DLContext ctx;
+  int ndim;
+  DLDataType dtype;
+  int64_t* shape;
+  int64_t* strides;
+  uint64_t byte_offset;
+} DLConstTensor;
+#ifdef __cplusplus
+}
+#endif
+
+namespace tc {
+/**
+ * C++ unique_ptr abstraction on top of DLTensor/DLConstTensor
+ */
+struct DLTensorDeleter {
+  template <typename DLTensorType>
+  void operator()(const DLTensorType* t) {
+    if (t->shape) {
+      delete[] t->shape;
```

Comment:
Do we have a utility for allocating `DLTensors` with `new[]`?  There's a risk of freeing the memory allocated with `malloc` in C by `delete[]` in C++, which has undefined behavior.

Reply:
yes we do and it is the only way to create DLTensorUPtr for exactly that reason.